# 植物分类

## 基本情况说明：

选取相关的植物，尽量覆盖整个植物系统发育的情况：

- Selaginella为卷柏属，Lycopodium为石松属，之前都归在Pteridophyta（蕨类植物门）之下。后来将蕨类植物中的卷柏分出，作为石松类植物。目前石松类植物只有两个基因组，均为卷柏属下的植物，其中一个并不含有注释文件。目前Lycopodiopsida（石松纲）下有3个部分：Isoetales/quillworts（水韭目），Lycopodiales（石松目），Selaginellales/spike mosses（卷柏目）。
- 蕨类（Polypodiidae/Fern）：目前主要包括3个已测序的目
  - 桫椤目（Cyatheales）
  - 水龙骨目（Polypodiales），两个测序的物种都在凤尾蕨科（Pteridaceae）
  - 槐叶萍目（Salviniales），两个测序的物种都在槐叶萍科（Salviniaceae）
    - Adiantum capillus-veneris（铁线蕨）目前刚刚被测序完成并组装，且提供注释文件，目前尚未被数据库收录。
- 裸子植物（Gymnosperm）：目前主要包括4个主要的lineages
  - 苏铁（Cycads）
  - 银杏（Ginkgo）
  - 松柏类（Conifers）
  - 买麻藤（Gnetophyte）
- 黄藤（Daemonorops jenkinsiana）在NCBI中更改为Calamus jenkinsiana，两个名字均一样。

## 一些中文名称

- 藻类：
  - Chlamydomonas reinhardtii - 莱茵绿藻
  - Volvox carteri - 强壮团藻（多细胞绿藻，与莱茵绿藻closely related）
  - Chromochloris zofingiensis - 佐夫色绿球藻
  - Botryococcus braunii - 布朗氏葡萄藻
  - Mesostigma viride - 中眼藻（鞭毛单细胞轮藻 - 链形植物门最早分化出的轮藻基因组）
  - Chlorokybus atmophyticus - 绿叠球藻（多细胞 - 链形植物门最早分化出的轮藻基因组）
  - Chara braunii - 布氏轮藻
  - Klebsormidium flaccidum - 软克里藻
  - Penium margaritaceum - 玛格丽塔柱形鼓藻（双星藻纲）
  - Mesotaenium endlicherianum - 中带鼓藻（双星藻纲）
- Basal land plants 包括 liverworts/mosses/hornworts：
  - Sphagnum magellanicum - 中位泥炭藓
  - Sphagum recurvum (old: Sphagnum fallax) - 喙叶泥炭藓
  - Physcomitrium patens - 立碗藓类
  - Marchantia polymorpha - 地钱
  - Ceratodon purpureus - 角齿藓
- Lycopodiopsida - 石松类：
  - Selaginella moellendorffii - 江南卷柏
  - Isoetes taiwanensis - 台湾水韭
- Fern - 蕨类：
  - Alsophila spinulosa - 桫椤
  - Ceratopteris richardii - 美洲水蕨
  - Adiantum capillus - 铁线蕨
  - Azolla filiculoides - 细叶满江红
  - Salvinia cucullata - 勺叶槐叶萍
- Gymnosperm - 裸子植物
  - Cycas panzhihuaensis - 攀枝花苏铁 
  - Ginkgo biloba - 银杏
  - Conifers II：
    - Taxus wallichiana - 红豆杉
    - Taxus wallichiana var. chinensis (old: Taxus chinensis) - 红豆杉
    - Sequoiadendron giganteum - 巨杉（giant sequoia）
  - Conifers I：
    - Pinus taeda - 火炬松（loblolly pine）
    - Pinus tabuliformis - 油松（southern Chinese pine）
    - Picea abies - 欧洲云杉（Norway spruce）
    - Picea glauca - 白云杉（White spruce）
    - Abies alba - 银冷杉（silver fir）
  - Gnetidae（买麻藤纲）：
    - Welwitschia mirabilis - 百岁兰
    - Gnetum montanum - 买麻藤
- Angiosperm - 被子植物
  - 被子植物基部
    - Amborella trichopoda - 无油樟
    - Nymphaea colorata - 蓝星睡莲
    - Nymphaea thermarum - 侏儒卢旺达睡莲/小睡莲
  - 两个单独分支
    - Ceratophyllum demersum - 金鱼藻
    - Chloranthus sessilifolius - 四川金粟兰
    - Chloranthus spicatus - 金粟兰
  - 木兰类
    - Chimonanthus salicifolius - 柳叶蜡梅
    - Cinnamomum micranthum (old: Cinnamomum kanehirae) - 沉水樟
    - Persea americana - 鳄梨
    - Liriodendron chinense - 鹅掌楸
  - 单子叶
    - Alismatales - 泽泻目
      - Araceae - 天南星科
        - Lemna minor - 浮萍
        - Spirodela polyrhiza - 紫萍
      - Zosteraceae - 大叶藻科
        - Zostera marina - 大叶藻
        - Zostera muelleri - 南半球海草（暂定）
    - Asparagales - 天门冬目
      - Amaryllidaceae - 石蒜科
        - Allium sativum - 蒜
      - Asparagaceae - 天门冬科
        - Asparagus officinalis - 石刁柏（俗名：芦笋）
        - Asparagus setaceus - 文竹
      - Orchidaceae - 兰科
        - Apostasia shenzhenica - 深圳拟兰
        - Gastrodia elata - 天麻
        - Dendrobium catenatum - 黄石斛
        - Phalaenopsis aphrodite - 蝴蝶兰
        - Phalaenopsis equestris - 小兰屿蝴蝶兰
        - Vanilla planifolia - 香荚兰
    - Arecales - 棕榈目
      - Arecaceae - 棕榈科
        - Cocos nucifera - 椰子
        - Elaeis guineensis - 油棕
        - Calamus simplicifolius - 单叶省藤
        - Calamus jenkinsiana (old: Daemonorops jenkinsiana) - 黄藤
        - Phoenix dactylifera - 海枣
    - Poales - 禾本目
      - Bromeliaceae - 凤梨科
        - Ananas comosus var. bracteatus (old: Ananas bracteatus) - 红凤梨
        - Ananas comosus - 凤梨
      - Cyperaceae - 莎草科
        - Carex littledalei (old: Kobresia littledalei) - 康藏嵩草
      - Poaceae - 禾本科
        - Phyllostachys edulis - 毛竹
        - Bonia amplexicaulis - 芸香竹
        - Guadua angustifolia - 瓜多竹
        - Olyra latifolia - 莪莉竹
        - Leersia perrieri 马达加斯加稻属，目前稻的外类群
        - Oryza barthii - 巴蒂稻
        - Oryza brachyantha - 短花稻
        - Oryza longistaminata - 长药稻
        - Oryza rufipogon - 野生稻
        - Puccinellia tenuiflora - 星星草
        - Lolium perenne - 黑麦草
        - Brachypodium distachyon - 二穗短柄草
        - Hordeum vulgare - 大麦
        - Aegilops tauschii - 节节麦
        - Triticum aestivum - 小麦
        - Triticum turgidum - 圆锥小麦
        - Eleusine coracana - 穇
        - Oropetium thomaeum 一种耐旱草类，尚无英文名，PacBio三代
        - Eragrostis curvula - 弯叶画眉草
        - Eragrostis tef - 苔麸
        - Zoysia japonica - 结缕草
        - Miscanthus sinensis - 芒
        - Saccharum spontaneum - 甜根子草
        - Sorghum bicolor - 高粱
        - Zea mays - 玉蜀黍（玉米）
        - Setaria italica - 粱
        - Setaria viridis - 狗尾草
        - Panicum virgatum - 柳枝稷
    - Zingiberales - 姜目
      - Musaceae - 芭蕉科
        - Ensete ventricosum - 粗柄象腿蕉
        - Musa schizocarpa 未知
  - 真双子叶基部
    - Ranunculales - 毛茛目
      - Circaeasteraceae - 星叶草科
        - Kingdonia uniflora - 独叶草
      - Papaveraceae - 罂粟科
        - Eschscholzia californica - 花菱草
        - Macleaya cordata - 博落回
        - Papaver somniferum - 罂粟
      - Ranunculaceae - 毛茛科
        - Aquilegia coerulea 耧斗菜的一种
    - Proteales - 山龙眼目
      - Nelumbonaceae - 莲科
        - Nelumbo nucifera - 莲
      - Proteaceae - 山龙眼科
        - Macadamia integrifolia - 澳洲坚果/夏威夷果
    - Trochodendrales - 昆栏树目
      - Trochodendraceae - 昆栏树科
        - Tetracentron sinense - 水青树
  - 核心真双子叶
    - Saxifragales - 虎耳草目
      - Crassulaceae - 景天科
        - Kalanchoe fedtschenkoi - 玉吊钟
    - Vitales - 葡萄目
      - Vitaceae - 葡萄科
        - Vitis vinifera - 葡萄
    - 蔷薇分支豆类
      - Celastrales - 卫矛目
        - Celastraceae - 卫矛科
          - Tripterygium wilfordii - 雷公藤
      - Oxalidales - 酢浆草目
        - Cephalotaceae - 土瓶草科
          - Cephalotus follicularis - 土瓶草
        - Oxalidaceae - 酢浆草科
          - Averrhoa carambola - 阳桃（杨桃）
      - Malpighiales - 金虎尾目
        - Euphorbiaceae - 大戟科
          - Ricinus communis - 蓖麻
          - Jatropha curcas - 麻风树
          - Manihot esculenta - 木薯
          - Hevea brasiliensis - 橡胶树
        - Linaceae - 亚麻科
          - Linum usitatissimum - 亚麻
        - Passifloraceae - 西番莲科
          - Passiflora edulis - 鸡蛋果（百香果）
        - Rhizophoraceae - 红树科
          - Kandelia obovata - 秋茄树
        - Salicaceae - 杨柳科
          - Populus alba - 银白杨
          - Populus pruinosa - 灰胡杨
          - Populus trichocarpa - 毛果杨
          - Salix brachista - 小垫柳
      - Fabales - 豆目
        - Fabaceae - 豆科
          - Chamaecrista fasciculata
          - Faidherbia albida - 环荚合欢
          - Mimosa pudica - 含羞草
          - Cercis canadensis - 加拿大紫荆
          - Arachis duranensis - 蔓花生
          - Arachis hypogaea - 落花生
          - Cajanus cajan - 木豆
          - Cicer arietinum - 鹰嘴豆
          - Glycine soja - 野大豆
          - Lablab purpureus - 扁豆
          - Lotus japonicus (Lotus corniculatus subsp. japonicus) - 光叶百脉根
          - Lupinus albus - 白羽扇豆
          - Lupinus angustifolius - 狭叶羽扇豆
          - Phaseolus lunatus - 棉豆
          - Phaseolus vulgaris - 菜豆
          - Pisum sativum - 豌豆
          - Spatholobus suberectus - 密花豆
          - Trifolium pratense - 红车轴草
          - Trifolium subterraneum - 地车轴草
          - Vigna angularis - 赤豆
          - Vigna radiata - 绿豆
          - Vigna unguiculata - 豇豆
      - Rosales - 蔷薇目
        - Cannabaceae - 大麻科
          - Cannabis sativa - 大麻
        - Moraceae - 桑科
          - Artocarpus altilis - 面包树
          - Artocarpus heterophyllus - 波罗蜜
        - Rosaceae - 蔷薇科
          - Prunus avium - 欧洲甜樱桃
          - Prunus mume - 梅
          - Prunus persica - 桃
          - Pyrus betulifolia - 杜梨
          - Pyrus communis - 西洋梨
          - Eriobotrya japonica - 枇杷
          - Dryas drummondii - 亮黄仙女木
          - Fragaria nilgerrensis - 黄毛草莓
          - Fragaria vesca - 野草莓
          - Rosa chinensis - 月季花
      - Cucurbitales - 葫芦目
        - Begoniaceae - 秋海棠科
          - Begonia fuchsioides - 倒挂金钟秋海棠
        - Cucurbitaceae - 葫芦科
          - Cucumis melo - 甜瓜
          - Cucumis sativus - 黄瓜
          - Lagenaria siceraria - 葫芦
          - Cucurbita argyrosperma - 墨西哥南瓜
          - Cucurbita maxima - 笋瓜
          - Cucurbita moschata - 南瓜
          - Cucurbita pepo - 西葫芦
        - Datiscaceae - 野麻科
          - Datisca glomerata - 头序野麻
      - Fagales - 壳斗目
        - Betulaceae - 桦木科
          - Alnus glutinosa - 欧洲桤木
        - Casuarinaceae - 木麻黄科
          - Casuarina glauca - 粗枝木麻黄
          - Castanea mollissima - 栗
        - Fagaceae - 壳斗科
          - Quercus suber - 欧洲栓皮栎
        - Juglandaceae - 胡桃科
          - Juglans nigra - 黑胡桃
          - Juglans regia - 胡桃
    - 蔷薇分支锦葵类
      - Myrtales - 桃金娘目
        - Lythraceae - 千屈菜科
          - Punica granatum - 石榴
        - Myrtaceae - 桃金娘科 
          - Eucalyptus grandis - 大桉
      - Sapindales - 无患子目
        - Anacardiaceae - 漆树科
          - Mangifera indica - 杧果
          - Pistacia vera - 阿月浑子（开心果）
          - Sclerocarya birrea - 象李
        - Rutaceae - 芸香科
          - Citrus clementina - 克莱门氏小柑橘
          - Citrus maxima (old: Citrus grandis) - 柚
          - Citrus cavaleriei (old: Citrus ichangensis) - 宜昌橙
          - Citrus medica - 香橼
          - Citrus sinensis - 甜橙
        - Sapindaceae - 无患子科
          - Dimocarpus longan - 龙眼
          - Xanthoceras sorbifolium - 文冠果
      - Malvales - 锦葵目
        - Malvaceae - 锦葵科
          - Bombax ceiba - 木棉
          - Theobroma cacao - 可可
          - Corchorus capsularis - 黄麻
          - Corchorus olitorius - 长蒴黄麻
          - Gossypium darwinii - 达尔文棉
          - Gossypium herbaceum - 草棉
          - Gossypium hirsutum - 陆地棉
          - Gossypium mustelinum - 黄褐棉
          - Gossypium raimondii - 美洲棉
          - Hibiscus syriacus - 木槿
        - Thymelaeaceae - 瑞香科
          - Aquilaria sinensis - 土沉香
      - Brassicales - 十字花目
        - Brassicaceae - 十字花科
          - Aethionema arabicum - 小亚细亚岩芥
          - Brassica rapa - 蔓菁
        - Moringaceae - 辣木科
          - Moringa oleifera - 辣木
        - Caricaceae - 番木瓜科
          - Carica papaya - 番木瓜
    - Caryophyllales - 石竹目
      - Amaranthaceae - 苋科
        - Amaranthus hypochondriacus - 千穗谷
      - Chenopodiaceae - 藜科
        - Beta vulgaris - 甜菜
      - Polygonaceae - 蓼科
        - Fagopyrum esculentum - 荞麦
    - 菊类分支
      - Cornales - 山茱萸目
        - Nyssaceae - 蓝果树科
          - Davidia involucrata - 珙桐
      - Ericales - 杜鹃花目
        - Actinidiaceae - 猕猴桃科
          - Actinidia chinensis - 中华猕猴桃
          - Actinidia eriantha - 毛花猕猴桃
        - Ebenaceae - 柿树科
          - Diospyros lotus - 君迁子
        - Ericaceae - 杜鹃花科
          - Rhododendron delavayi - 马缨杜鹃
      - 菊分支唇形类
        - Garryales - 丝樱花目
          - Eucommiaceae - 杜仲科
            - Eucommia ulmoides - 杜仲
        - Solanales - 茄目
          - Convolvulaceae - 旋花科（包含菟丝子科）
            - Cuscuta australis - 南方菟丝子
            - Cuscuta campestris - 原野菟丝子
            - Ipomoea triloba - 三裂叶薯（红花野牵牛）
          - Solanaceae - 茄科
            - Nicotiana attenuata - 渐狭叶烟草
            - Capsicum annuum - 辣椒
            - Solanum chilense - 智利番茄
            - Solanum lycopersicum - 番茄（小番茄）
            - Solanum pennellii - 彭氏番茄
            - Solanum tuberosum - 马铃薯
        - Gentianales - 龙胆目
          - Apocynaceae - 夹竹桃科
            - Calotropis gigantea - 牛角瓜
          - Rubiaceae - 茜草科
            - Coffea canephora - 中粒咖啡
        - Lamiales - 唇形目
          - Bignoniaceae - 紫葳科
            - Handroanthus impetiginosus - 粉红钟花
          - Gesneriaceae - 苦苣苔科
            - Boea hygrometrica - 旋蒴苣苔
          - Lamiaceae - 唇形科
            - Salvia bowleyana - 南丹参
          - Lentibulariaceae - 狸藻科
            - Utricularia gibba - 少花狸藻（丝叶狸藻）
          - Oleaceae - 木犀科
            - Olea europaea - 木樨榄（油橄榄）
          - Phrymaceae - 透骨草科
            - Mimulus guttatus - 多斑沟酸浆
      - 菊分支桔梗类
        - Asterales - 菊目
          - Astereae - 菊科
            - Chrysanthemum indicum - 野菊
            - Chrysanthemum lavandulifolium (old: Chrysanthemum seticuspe) - 甘菊
            - Erigeron breviscapus - 短莛飞蓬
            - Mikania micrantha - 微甘菊
            - Helianthus annuus - 向日葵
            - Cynara cardunculus - 刺苞菜蓟
            - Lactuca sativa - 莴苣
        - Apiales - 伞形目
          - Apiaceae - 伞形科
            - Apium graveolens - 旱芹（芹菜）
            - Coriandrum sativum - 芫荽（香菜）
        - Dipsacales - 川续断目
          - Caprifoliaceae - 忍冬科
            - Lonicera japonica - 忍冬

## Preparation

Visit [Anaconda](https://www.anaconda.com/) to download install script.

- BUSCO

Visit [BUSCO userguide](https://busco.ezlab.org/busco_userguide.html#conda-package) for database.

```bash
# install anaconda
bash Anaconda3-2022.05-Linux-x86_64.sh
# enter to allow and complete installation
# start linux again, directly activating base

# avoid directly activating base when start linux
conda config --set auto_activate_base false
# conda install busco
conda create -n genome -c conda-forge -c bioconda busco=5.4.2
# activate env
conda activate genome
```

- OrthoFinder

```bash
conda config --add channels defaults
conda config --add channels bioconda
conda config --add channels conda-forge

# install orthofinder
conda install orthofinder
# primary_transcript.py to get the proteins from the primary transcripts
```

- HMMER

```bash
mkdir -p ~/data/symbio/HMM/PFAM
cd ~/data/symbio/HMM/PFAM

for basename in Pfam-A.hmm Pfam-A.hmm.dat active_site.dat; do
    wget -N --content-disposition ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/${basename}.gz
    wget -N --content-disposition ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/${basename}.gz
    wget -N --content-disposition ftp://ftp.ebi.ac.uk/pub/databases/Pfam/releases/Pfam35.0/${basename}.gz
done

for basename in Pfam-A.hmm Pfam-A.hmm.dat active_site.dat; do
    echo "==> ${basename}"
    gzip -dcf ${basename}.gz > ${basename}
done

hmmpress Pfam-A.hmm
```

- Other software

```bash
brew install dos2unix
```

## CDS or protein selection

### Genomes selection and CDS renaming

There are something to declare: 

- Multiple transcripts were not excluded here, which means different types of the species transcripts were not specified. Maybe the primary transcripts would be adopted afterwards.
- All genomes were put into directories named by `species`, then moved into directory `GENOMES`. Every directory basically had four files including genome, gff, cds and pep.
- Download directly from URL (no records in database)
  - fagopyrum_esculentum
  - allium_sativum
  - elaeis_guineensis
  - passiflora_edulis
  - daemonorops_jenkinsiana
  - sphagnum_magellanicum
  - adiantum_capillus-veneris (no longest gff)
  - ceratopteris_richardii, marchantia_paleacea
  - carica_papaya
  - taxus_chinensis
  - sphagnum_fallax
- Discard
  - punica_granatum (gff has problem)
  - picea_abies (ftp cannot be accessed)
  - phoenix_dactylifera

```bash
mkdir ~/data/symbio/info
cd ~/data/symbio/info

# manually pick some genomes into taxo_genom.xlsx

perl ~/Scripts/fig_table/xlsx2csv.pl -f taxo_genom.xlsx |
    tsv-select -d, -f 1 |
    sed 1d |
    perl -nl -e '$_ = lc $_;print $_' \
    > genome.lst

wc -l genome.lst
#148 genome.lst

# process genomes
cd ~/data/symbio/GENOMES

for dir in $(ls)
do
    echo "==>{dir}"
    faops size ${dir}/genome.cds |
        cut -f 1 |
        awk -v spe=${dir} '{print ($0"\t"spe_$0)}' \
        > tmp
    faops replace ${dir}/genome.cds tmp ../CDS/${dir}.cds.fa
    rm tmp
done

# there would have a question: faops size will automatically stop at the first space
# so some cds files would be named incorrectly
# check whether changed successfully
cd ~/data/symbio

cat info/genome.lst |
    parallel -j 6 --keep-order '
        echo "==>{}"
        faops size CDS/{}.cds.fa | cut -f 1 | head -n 1
    '
# manually check those cds done wrong

# manually change those species with problems
# following is an example for dealing with the `.||.||.` type cds
cd ~/data/symbio/GENOMES/<species>

# <species> means you should replace to the specific species
cat genome.cds |
    perl -nle '
        print uc $_ if /^[AGCTN]/i;
        print "><species>_$1" if /^>.+\|(.+?)\|\|-*\d\|\|CDS/;
    ' \
    > ../../CDS/<species>.cds.fa 

# finally check the renamed cds numbers
cat info/genome.lst |
    parallel -j 6 --keep-order '
        echo "==>{}"
        CDS=$(cat CDS/{}.cds.fa | faops size stdin | wc -l)
        GEN=$(cat GENOMES/{}/genome.cds | faops size stdin | wc -l)
        if [[ $CDS == $GEN ]]; then
            echo check ok
        else
            echo check wrong
        fi
    '
# nothing wrong
```

### Protein processing

CDS always failed in grouping orthogroups correctly, so I chose protein files.

```bash
mkdir ~/data/symbio/PROTEIN
cd ~/data/symbio

cat info/genome.lst |
    parallel -j 6 --keep-order '
        if [ ! -f GENOMES/{}/genome.pep ]; then
            echo "==>{} without protein"
        else
            cp GENOMES/{}/genome.pep PROTEIN/{}.pep.fa
        fi
    '
# remove all species without pro seq in the genome.lst

conda activate genome

for file in $(ls PROTEIN)
do
    echo "==> ${file}"
    python /home/j/anaconda3/envs/genome/bin/primary_transcript.py PROTEIN/${file}
    echo
done

cd PROTEIN
ls primary_transcripts | wc -l
#148

mv primary_transcripts/ ..
cd ..

# make sure which file should change seq_id
# check seq_id could not be change
for file in $(ls primary_transcripts)
do
    id=$(cat primary_transcripts/${file} | grep '^>' | head -n 1)
    result=$(echo $id | grep '\s')
    if [[ $result != "" ]]; then
        echo "==>${file} name should change"
        echo $id
        echo
    fi
done
#persea_americana
#lemna_minor
#ceratophyllum_demersum
#cat primary_transcripts/<species>.pep.fa |
#    perl -nle '
#        print ">$1" if /^>.+?\|\|.+?\|\|.+?\|\|.+?\|\|(.+?)\|/;
#        print if /^[A-Z]/i;
#    ' > tmp && mv tmp primary_transcripts/<species>.pep.fa

for file in $(ls primary_transcripts)
do
    id=$(cat primary_transcripts/${file} | grep '^>' | head -n 1)
    result=$(echo $id | grep '\s')
    if [[ $result != "" ]]; then    
        cat primary_transcripts/${file} | perl -nle '
            print ">$1" if /^>(.+?)\s/;
            print if /^[A-Z]/i;
        ' \
        > tmp && mv tmp primary_transcripts/${file}
    fi
done
# run above code for sure again

# add species name before the id
for name in $(ls primary_transcripts | perl -p -e 's/\..+$//')
do
    faops size primary_transcripts/${name}.pep.fa |
        cut -f 1 |
        awk -v name=${name} '{print ($0"\t"name"_"$0)}' \
        > name.tmp
    faops replace primary_transcripts/${name}.pep.fa name.tmp tmp.fa \
    && mv tmp.fa primary_transcripts/${name}.pep.fa
done

# some protein files using . for termination codon
# change them to * for satisfying diamond
for file in $(ls primary_transcripts)
do
    SEQ=$(faops some -l 0 primary_transcripts/${file} \
            <(faops size primary_transcripts/${file}) \
            stdout | grep -v '^>')
    result=$(echo $SEQ | grep '\.')
    if [[ $result != "" ]]; then
        echo "==>${file} end is ."
    fi
done

# manually change them
for file in $(ls primary_transcripts)
do
    cat primary_transcripts/${file} |
        perl -nle 'if(/^>/){print;}else{$_ =~ s/\./\*/g; print $_;}' > tmp \
        && mv tmp primary_transcripts/${file}
done
# run above code for sure again

rsync -avP primary_transcripts/ name@ip:jyq/data/symbio/
```

## BUSCO

Running BUSCO for checking the genomes or assemblies sequencing quality.

```bash
mkdir ~/data/symbio/BUSCO
cd ~/data/symbio

# check busco of every genome
bash scripts/busco.sh
```

## OrthoFinder identifying gene orthogroups

### OrthoFinder processing

```bash
# do following to transmit data to server
#rsync -avP /home/j/data/symbio/CDS name@ip:jyq/data/

# connect to server
# ssh name@ip

cd ~/jyq/data/symbio

# orthofinder in conda base env
conda activate

orthofinder -f ./primary_transcripts -og
# orthofinder [options] -f <dir>
# -og: Stop after inferring orthogroups

# cp results to local
#rsync -avP name@ip:jyq/data/symbio/primary_transcripts/OrthoFinder/Results_Oct20/Orthogroups ~/data/symbio/

# If hard limit, h > r already, then you just need to increase the soft limit:
ulimit -n 22004

# check the limit
ulimit -Sn

# continue after orthogroups inferred
orthofinder -fg Results_Oct20/ -M msa -X
# -fg <dir>: Start OrthoFinder from pre-computed orthogroups in <dir>
# -M <txt>: Method for gene tree inference. Options 'dendroblast' & 'msa' [Default = dendroblast]
# -X: Don't add species names to sequence IDs
```

### OrthoFinder results

Acquire results from the workstation using rsync and put in the dir `Orthogroups` (command not shown).

```bash
cd ~/data/symbio/Orthogroups
mkdir groups

cat Orthogroups.tsv | cut -f 1 | sed 1d > ortho.lst

wc -l ortho.lst
#135923 ortho.lst
# Totally 135923 orthogroups were identified

# split each orthogroup into a tsv
cat ortho.lst |
    parallel -j 4 --keep-order '
        echo "==> {}"
        bash ../scripts/ortho_extract.sh {} ./groups/
    '

mkdir pro3
# extract 3 longest proteins
cat ortho.lst |
    parallel -j 16 --keep-order '
        echo "==> {}"
        bash ../scripts/pro_extract.sh groups/{}.csv ../primary_transcripts/ {}
        if [ -f {}.fa ]; then
            cat {}.fa |
                faops size stdin |
                sort -nk 2,2 -r |
                head -n 3 |
                cut -f 1 \
                > {}.lst
            faops some {}.fa {}.lst ./pro3/{}.fa
            rm {}.lst {}.fa
        else
            echo problem
        fi
    '
```

### Renaming and extracting

After OrthoFinder processing, those pep files contained `:` in their seq_ids will be converted to `_`. Protein extraction will fail if you do not change your seq_ids.

The pep file of ceratophyllum_demersum has predicted sequences generated from six frame translation, remove them manually (code not shown).

```bash
cd ~/data/symbio

rm info/rename_id.lst
# check those files whose seq_ids should be changed
for file in $(ls primary_transcripts)
do
    if grep -i -q ':' primary_transcripts/${file}; then
        echo "${file}" >> info/rename_id.lst
    fi
done
#anthoceros_angustus.pep.fa
#bonia_amplexicaulis.pep.fa
#calamus_simplicifolius.pep.fa
#cocos_nucifera.pep.fa
#daemonorops_jenkinsiana.pep.fa
#mesotaenium_endlicherianum.pep.fa
#olyra_latifolia.pep.fa
#phalaenopsis_equestris.pep.fa
#phyllostachys_edulis.pep.fa
#rhododendron_delavayi.pep.fa
#welwitschia_mirabilis.pep.fa

for file in $(cat info/rename_id.lst)
do
    cat primary_transcripts/${file} |
        perl -p -e 's/:/_/g' \
        > tmp && mv tmp primary_transcripts/${file}
done

# run again to check whether renaming succeed
for file in $(ls primary_transcripts)
do
    if grep -i -q ':' primary_transcripts/${file}; then
        echo "${file}"
    fi
done
# OK
```

### Orthogroups RLK identifying

- Identify domains from each protein seq via `hmmscan`

```bash
cd ~/data/symbio
mkdir -p ~/data/symbio/DOMAIN/pfam

cat info/genome.lst |
    parallel -j 6 --keep-order '
        echo {}
        hmmscan --cpu 2 -E 0.1 --domE 0.1 -o DOMAIN/pfam/{}.txt \
            ./HMM/PFAM/Pfam-A.hmm primary_transcripts/{}.pep.fa
    '

ls DOMAIN/pfam/*.txt |
    parallel -j 6 --keep-order '
        echo "==> {/.}"
        perl scripts/hmm_results.pl -i {} \
        > DOMAIN/pfam/{/.}.pfam.tsv
    '
```

- Extract kinase domain

```bash
cd ~/data/symbio/DOMAIN

rm all_domains.tsv
# all domains were extracted
for file in $(ls pfam/*.pfam.tsv)
do
    cat ${file} |
        tsv-select -H -f Domain |
        sed 1d \
        >> all_domains.tsv
done

# basic info of domains identified
cat all_domains.tsv |
    tsv-summarize -g 1 --count |
    sort -r -nk 2,2 \
    > tmp && mv tmp all_domains.tsv

cat all_domains.tsv | wc -l
#19622
# all domains identified, including those repeated domains from a location
# this will be considered later

# extract all domains of pkinase with cutoff E-value <= 1e-4
# be aware of pk may contained domains not only pkinase
ls pfam/*.pfam.tsv |
    perl -p -e 's/\.pfam\.tsv$//' |
    parallel -j 12 --keep-order '
        echo "==> {/}"
        cat {}.pfam.tsv |
            tsv-filter -H --iregex Domain:pk --le E_value:"1e-4" |
            sed 1d |
            tsv-select -f 1 |
            tsv-uniq \
            > KD/{/}.lst
        cat {}.pfam.tsv |
            sed 1d |
            tsv-join -f KD/{/}.lst -k 1 |
            tsv-select -f 1,2,4,5,3 \
            > KD/{/}.tsv
        rm KD/{/}.lst
    '

rm all_KD.tsv
for file in $(ls KD/*.tsv)
do
    cat ${file} |
        tsv-select -f 2 |
        tsv-filter --iregex 1:pk \
        >> all_KD.tsv
done

cat all_KD.tsv |
    tsv-summarize -g 1 --count |
    sort -r -nk 2,2 \
    > tmp && mv tmp all_KD.tsv

rm all_KD_species.tsv
# all proteins with the pk domain among species
ls KD/*.tsv |
    parallel -j 12 --keep-order '
        cat {} |
            cut -f 1 |
            tsv-uniq |
            wc -l |
            awk -v spe={/.} '\''{print (spe"\t"$0)}'\'' \
            >> all_KD_species.tsv
    '

# all proteins
cat all_KD_species.tsv |
    cut -f 2 |
    tr '\n' '+' |
    sed 's/+$/\n/' |
    bc
#204137

# extract all protein sequences contained the kinase domain
mkdir -p ~/data/symbio/DOMAIN/seq_KD

ls KD/*.tsv |
    parallel -j 12 --keep-order '
        echo "==> {/.}"
        faops some -l 0 \
            ../primary_transcripts/{/.}.pep.fa \
            <(cat {} | cut -f 1 | tsv-uniq) \
            ./seq_KD/{/.}.fa
    '

# check extraction
wc -l ./seq_KD/*.fa | grep 'total' | perl -p -e 's/\s+(\d+).+$/$1\/2/' | bc
#204137
# extraction complete
```

- `TMHMM2` for transmembrane domain identification

Manually upload files to [TMHMM](https://services.healthtech.dtu.dk/service.php?TMHMM-2.0) and copy results into a tsv file.

Although `selenium` could deal with this automatically, but there are always problems that I cannot solve. But still, a script will be saved in [scripts](https://github.com/jdasfd/symbio/tree/main/scripts). It will be completed later (maybe). XD

```bash
mkdir -p ~/data/symbio/DOMAIN/TMD
# manually copy all results into a tsv

mkdir ~/data/symbio/DOMAIN/results
dos2unix *.tsv

for file in $(ls)
do
    echo "==> ${file}"
    perl ../../scripts/tmhmm_result.pl -i ${file} > ../results/${file}
done
```

## RNA-seq of AM symbiosis

```bash
mkdir ~/data/symbio/sra
cd ~/data/symbio/sra


```

```bash
for file in $(ls)
do
    hmmscan --cpu 8 -E 1e-4 --domE 1e-4 -o ${file}.txt \
    ../../symbio/HMM/PFAM/Pfam-A.hmm ${file}
done
```
